components/ILIAS/Refinery/src/Encode/Transformation/HTMLAttributeValue.php

```diff
@@ -0,0 +1,107 @@
+ * The given string is encoded so it can be used in any HTML Tag regardless if the value is enclosed in double quotes ("), single quotes (') or no quotes at all (e.g. <div class=ENCODED_VALUE></div>).
+ * Please see for more information of the syntax of HTML attributes: https://html.spec.whatwg.org/multipage/syntax.html#attributes-2
+ * Please see https://html.spec.whatwg.org/multipage/dom.html#phrasing-content for a list of all allowed characters and escape sequences.
+ */
+class HTMLAttributeValue implements Transformation
+{
+    use DeriveInvokeFromTransform;
+    use DeriveApplyToFromTransform;
+
+    public function transform($from)
+    {
+        return $this->encode($from);
+    }
+
+    private function encode(string $from): string
+    {
+        return preg_replace_callback(
+            '/[^a-z0-9,._-]/iSu',
+            fn(array $m): string => $this->replace($m[0]),
+            $from
+        ) ?? throw new ValueError('Invalid UTF-8 string given.');
+    }
+
+    private function replace(string $utf8_char): string
+    {
+        $codepoint = $this->utf8CharacterToCodepoint($utf8_char);
+
+        // All unicode control characters besides white space codepoints as well as noncharacters are not allowed in HTML attributes.
+        if ($this->isNonPrintableControl($utf8_char, $codepoint) || $this->isNonCharacter($codepoint)) {
+            return '&#xFFFD;'; // Unicode "replacement character", indicating a non printable character.
+        }
+
+        return match ($codepoint) {
+            34 => '&quot;',
+            38 => '&amp;',
+            60 => '&lt;',
+            62 => '&gt;',
+            default => sprintf($codepoint > 255 ? '&#x%04X;' : '&#x%02X;', $codepoint),
+        };
+    }
+
+    /**
+     * Decodes a given UTF-8 character (which may be multibyte) to a Unicode codepoint.
+     * E.g. The multibyte character "€", encoded in UTF-8 as the byte sequence 0xE2 0x82 0xAC,
+     * will return the integer 0x20AC, which is the corresponding unicode codepoint.
+     */
+    private function utf8CharacterToCodepoint(string $utf8_char): int
+    {
+        // UTF-32 encodes Unicode codepoints as itself. BE stands for Big Endian.
+        return hexdec(bin2hex(strlen($utf8_char) > 1 ? (mb_convert_encoding($utf8_char, 'UTF-32BE', 'UTF-8')) : $utf8_char));
+    }
+
+    /**
+     * The unicode range for control codepoints is from U+0000 NULL to U+001F (inclusive) and from U+007F to U+009F (inclusive).
+     * The range U+0000 NULL to U+001F and the codepoint U+007F are all one byte long and are covered by the PHP function ctype_cntrl.
+     *
+     * See https://infra.spec.whatwg.org/#control for more information.
+     */
+    private function isNonPrintableControl(string $utf8_char, int $codepoint): bool
+    {
+        return strlen($utf8_char) === 1 ?
+            ctype_cntrl($utf8_char) && !ctype_space($utf8_char) :
+            0x1F <= $codepoint && $codepoint <= 0x9F;
```

Comment:
Nitpick: the `0x1F <= $codepoint && ` part is a bit confusing after you have read the function comment. It does not matter, but it will always be true; we could remove it. Or replace `0x1F` with `0x7F` (which would make the check less strict, but still does not matter).

If I’m mistaken, please enlighten me; if not, consider dropping the part from the code. Thanks. :smiley:

Reply:
Is removed